usr/lib/ica_s390_stdll/ica_specific.c

```diff
@@ -3124,6 +3125,364 @@ static CK_RV os_specific_rsa_decrypt(STDLL_TokData_t *tokdata,
+            buffer = (CK_BYTE *)malloc(AES_BLOCK_SIZE);
+            if (!buffer) {
+                TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                rc = CKR_HOST_MEMORY;
+                goto done;
+            }
+            memcpy(buffer, context->data, context->len);
+
+            /* Perform final update with rest of data to calculate final tag */
+            rc = ica_aes_gcm_kma_update(buffer, out_data, context->len,
+                                        NULL, 0, 1,1, gcm_ctx);
+            if (rc != 0) {
+                TRACE_ERROR("ica_aes_gcm_kma_update failed, rc=%ld\n",rc);
+                rc = CKR_FUNCTION_FAILED;
+                goto done;
+            }
+
+            *out_data_len = context->len + tag_data_len;
+        } else {
+            /* Perform final update without data to calculate final tag */
+            rc = ica_aes_gcm_kma_update(NULL, NULL, 0, NULL, 0, 1, 1, gcm_ctx);
+            *out_data_len = tag_data_len;
+        }
+
+        TRACE_DEVEL("GCM Final: context->len=%ld, tag_data_len=%ld, "
+                    "out_data_len=%ld\n",
+                    context->len, tag_data_len, *out_data_len);
+
+        rc = ica_aes_gcm_kma_get_tag(tmp_tag, tag_data_len, gcm_ctx);
+        if (rc != 0) {
+            TRACE_ERROR("ica_aes_gcm_kma_get_tag failed, rc=%ld.\n", rc);
+            rc = CKR_FUNCTION_FAILED;
+            goto done;
+        }
+
+        memcpy(out_data + context->len, tmp_tag, tag_data_len);
+
+    } else {
+        /* decrypt */
+        if (context->len > tag_data_len) {
+            buffer = (CK_BYTE *)malloc(AES_BLOCK_SIZE);
+            if (!buffer) {
+                TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+                rc = CKR_HOST_MEMORY;
+                goto done;
+            }
+            memcpy(buffer, context->data, context->len - tag_data_len);
+
+            /* Perform final update with rest of data to calculate final tag */
+            rc = ica_aes_gcm_kma_update(buffer, out_data,
+                                        context->len - tag_data_len,
+                                        NULL, 0, 1, 1, gcm_ctx);
+            if (rc != 0) {
+                TRACE_ERROR("ica_aes_gcm_kma_update failed, rc=%ld.\n",rc);
+                rc = CKR_FUNCTION_FAILED;
+                goto done;
+            }
+            (*out_data_len) = context->len - tag_data_len;
+
+        } else if (context->len == tag_data_len) {
+            /* remaining data are tag data */
+            *out_data_len = 0;
+
+            /* Perform final update without data to calculate final tag */
+            rc = ica_aes_gcm_kma_update(NULL, NULL, 0, NULL, 0, 1, 1, gcm_ctx);
+        } else {                /* (context->len < tag_data_len) */
+            TRACE_ERROR("Incoming data are not consistent.\n");
+            rc = CKR_DATA_INVALID;
+            goto done;
+        }
+
+        final_tag_data = context->data + context->len - tag_data_len;
```

Comment:
Is it guaranteed that this does not overrun the `context->data` buffer? `context->data` is `2*AES_BLOCK_SIZE` in size. `context->len` is probably not larger than one `AES_BLOCK_SIZE`, so one `AES_BLOCK_SIZE `is left. Is this sufficient for `tag_data_len` ?

Reply:
Yes, valid tag sizes are from 4 bytes up to 16 bytes.